src/iceberg/json_internal.cc

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "iceberg/json_internal.h"
+
+#include <format>
+
+#include <nlohmann/json.hpp>
+
+#include "iceberg/sort_order.h"
+#include "iceberg/transform.h"
+#include "iceberg/util/formatter.h"
+
+namespace iceberg {
+
+namespace {
+
+constexpr std::string_view kTransform = "transform";
+constexpr std::string_view kSourceId = "source-id";
+constexpr std::string_view kDirection = "direction";
+constexpr std::string_view kNullOrder = "null-order";
+
+constexpr std::string_view kOrderId = "order-id";
+constexpr std::string_view kFields = "fields";
+
+// --- helper for safe JSON extraction ---
+template <typename T>
+expected<T, Error> GetJsonValue(const nlohmann::json& json, std::string_view key) {
+  if (!json.contains(key)) {
+    return unexpected<Error>({.kind = ErrorKind::kInvalidArgument,
+                              .message = "Missing key: " + std::string(key)});
+  }
+  try {
+    return json.at(key).get<T>();
+  } catch (const std::exception& ex) {
+    return unexpected<Error>({.kind = ErrorKind::kInvalidArgument,
+                              .message = std::string("Failed to parse key: ") +
+                                         key.data() + ", " + ex.what()});
+  }
+}
+
+#define TRY_ASSIGN(json_value, expr)                                    \
```

Comment:
We'll probably want to make a broader version of this macro. (Not necessarily in this PR)

FWIW, you could steal the Arrow version, which uses a counter to autogenerate the name, so it can be used like `ARROW_ASSIGN_OR_RAISE(auto value, ...)` or `ARROW_ASSIGN_OR_RAISE(foo.bar, ...)` too

Reply:
I did this exactly: https://github.com/apache/iceberg-cpp/pull/65/files#diff-d8a08fcbce5230d86b8b253508bd5d8dc6216a2d0f5049b58b6e7792f2243040